cirq/ops/linear_combinations.py

```diff
@@ -205,3 +207,124 @@ def extend(expansion: value.LinearDict[str],
             extended_expansion = extend(expansion, op.qubits, self.qubits)
             result += extended_expansion * coefficient
         return result
+
+
+def _is_linear_dict_of_unit_pauli_string(
+        linear_dict: value.LinearDict[UnitPauliStringT]) -> bool:
+    if not isinstance(linear_dict, value.LinearDict):
+        return False
+    for k in linear_dict.keys():
+        if not isinstance(k, frozenset):
+            return False
+        for qid, pauli in k:
+            if not isinstance(qid, raw_types.Qid):
+                return False
+            if not isinstance(pauli, pauli_gates.Pauli):
+                return False
+
+    return True
+
+@value.value_equality(approximate=True)
+class PauliSum:
+    """Represents operator defined by linear combination of PauliStrings.
+
+    Since PauliStrings store their own coefficients, this class
+    does not implement the LinearDict interface. Instead, you can
+    add and subtract terms and then iterate over the resulting
+    (simplified) expression.
+
+    Under the hood, this class is backed by a LinearDict with coefficient-less
+    PauliStrings as keys. PauliStrings are reconstructed on-the-fly during
+    iteration.
+
+    """
+
+    def __init__(self, linear_dict: value.LinearDict[UnitPauliStringT]):
+        if not _is_linear_dict_of_unit_pauli_string(linear_dict):
+            raise ValueError(
+                "PauliSum constructor takes a LinearDict[UnitPauliStringT]. "
+                "Consider using PauliSum.from_pauli_strings() or adding and "
+                "subtracting PauliStrings")
+        self._linear_dict = linear_dict
+
+    def _value_equality_values_(self):
+        return self._linear_dict
+
+    @classmethod
+    def from_pauli_strings(cls, terms: List[PauliString]) -> 'PauliSum':
```

Comment:
Update type annotation to Union[PauliString, List[pauliString]]

Reply:
done